PocketKit/Sources/Sync/Space.swift

```diff
@@ -272,57 +324,112 @@ public class Space {
-        )
+    func batchDeleteOrphanedSlates() throws {
+        let fetchRequest: NSFetchRequest<NSFetchRequestResult> = Slate.fetchRequest()
+        fetchRequest.predicate = NSPredicate(format: "slateLineup = NULL")
+        let deleteRequest = NSBatchDeleteRequest(fetchRequest: fetchRequest)
+        _ = try backgroundContext.performAndWait {
+            try backgroundContext.execute(deleteRequest)
+        }
     }
 
-    func backgroundObject<T: NSManagedObject>(with id: NSManagedObjectID) -> T? {
-        backgroundContext.performAndWait {
-            backgroundContext.object(with: id) as? T
+    /// Updates a `SlateLineup` from the specified remote object, on a child background context
+    /// - Parameter remote: the specified remote object
+    func updateLineup(from remote: SlateLineup.RemoteSlateLineup) throws {
+        let context = makeChildBackgroundContext()
+
+        context.performAndWait { [weak self] in
+            guard let self else { return }
+
+            let lineup = (try? self.fetchSlateLineup(byRemoteID: remote.id, context: context)) ??
+            SlateLineup(context: context, remoteID: remote.id, expermimentID: remote.experimentId, requestID: remote.requestId)
+
+            lineup.update(from: remote, in: self, context: context)
+        }
+        // save the child context
+        try context.performAndWait {
+            guard context.hasChanges else {
+                return
+            }
+            try context.save()
+            // then save the parent context
+            try save()
+            // then purge orphaned objects from the parent
+            try batchDeleteOrphanedSlates()
+            try batchDeleteOrphanedItems()
         }
     }
 
-    func viewObject<T: NSManagedObject>(with id: NSManagedObjectID) -> T? {
-        viewContext.performAndWait {
-            viewContext.object(with: id) as? T
+    /// Updates a `Slate` from the specified remote object, on a child background context
+    /// - Parameter remote: the specified remote object
+    func updateSlate(from remote: Slate.RemoteSlate) throws {
+        let context = makeChildBackgroundContext()
+        context.performAndWait { [weak self] in
+            guard let self else { return }
+
+            let slate = (try? self.fetchSlate(byRemoteID: remote.id)) ??
+            Slate(context: context, remoteID: remote.id, expermimentID: remote.experimentId, requestID: remote.requestId)
+            slate.update(from: remote, in: self, context: context)
         }
-    }
 
-    func backgroundRefresh(_ object: NSManagedObject, mergeChanges: Bool) {
-        backgroundContext.performAndWait {
-            backgroundContext.refresh(object, mergeChanges: mergeChanges)
+        try context.performAndWait {
+            guard context.hasChanges else {
+                return
+            }
+            // save the child context
+            try context.save()
+            // then save the parent context
+            try save()
```

Comment:
Remembering to save the parent context, is gonna be fun 😬

Reply:
I agree @cyndichin! I'd only postpone this to a future PR because it feels a but odd to have to pass one context while having the other one hardcoded. A bit more refactoring will come and, at that point, an helper method will come in handy for sure.